webcrypto/aes.go

```diff
@@ -151,3 +155,458 @@ func importAESKey(
+	// The AES-GCM specification recommends that the IV should be 96 bits long, and
+	// typically contains bits from a random number generator.
+	//
+	// Section 8.2 of the specification outlines methods for constructing IVs.
+	// Note that the IV does not have to be secret, just unique: so it is OK, for example, to
+	// transmit it in the clear alongside the encrypted message.
+	Iv []byte `json:"iv"`
+
+	// AdditionalData (an ArrayBuffer, a TypedArray, or a DataView) contains additional data that will
+	// not be encrypted but will be authenticated along with the encrypted data.
+	//
+	// If additionalData is given here then the same data must be given in the corresponding call
+	// to decrypt(): if the data given to the decrypt() call does not match the original data, the
+	// decryption will throw an exception.
+	// This gives you a way to authenticate associated data without having to encrypt it.
+	//
+	// The bit length of additionalData must be smaller than 2^64 - 1.
+	//
+	// The additionalData property is optional and may be omitted without compromising the
+	// security of the encryption operation.
+	AdditionalData []byte `json:"additionalData"`
+
+	// TagLength (a Number) determines the size in bits of the authentication tag generated in
+	// the encryption operation and used for authentication in the corresponding decryption.
+	//
+	// According to the Web Crypto specification this must have one of the
+	// following values: 32, 64, 96, 104, 112, 120, or 128.
+	// The AES-GCM specification recommends that it should be 96, 104, 112, 120 or 128, although
+	// 32 or 64 bits may be acceptable
+	// in some applications: Appendix C of the specification provides additional guidance here.
+	//
+	// tagLength is optional and defaults to 128 if it is not specified.
+	TagLength int `json:"tagLength"`
+}
+
+// Encrypt encrypts the given plaintext using the AES-GCM algorithm, and returns the ciphertext.
+// Implements the WebCryptoAPI's `encrypt` method's [specification] for the AES-GCM algorithm.
+//
+// [specification]: https://www.w3.org/TR/WebCryptoAPI/#aes-gcm
+func (agp *AesGcmParams) Encrypt(plaintext []byte, key CryptoKey) ([]byte, error) {
+	// 1.
+	// As described in section 8 of AES-GCM [NIST SP800-38D].
+	// [NIST SP800-38D] https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf
+	if len(plaintext) > maxAesGcmPlaintextLength {
+		return nil, NewError(0, OperationError, "plaintext length is too long")
+	}
+
+	// 2.
+	// As described in section 8 of AES-GCM [NIST SP800-38D].
+	// [NIST SP800-38D] https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf
+	//
+	// NOTE @oleiade: that the spec targets to support multiple IV lengths
+	// but go only supports 12 bytes IVs. We therefore are diverging from the
+	// spec here, and have adjusted the test suite accordingly.
+	if len(agp.Iv) != 12 {
+		return nil, NewError(0, OperationError, "iv length is not 12 bytes long")
```

Comment:
```suggestion
		return nil, NewError(0, NotSupportedErrror, "only 12 bytes long iv length are supported")
```
I hope it helps to understand better why